Update dependencies

abovedave opened the pull request:
* Upgrades all dependencies to latest versions, including typescript.
* Adds an `.nvmrc` file as we've done in other packages.
* Removes the old [`concat`](https://www.npmjs.com/package/concat) package which was causing ts errors in favour of a native solution.
* Couple of changes to `tsconfig.json` to resolve build errors.

abovedave committed:
add .nvmrc for v20

abovedave committed:
update assert to with for json imports

abovedave committed:
remove use of old contact package

abovedave committed:
update node

abovedave committed:
update all packages

abovedave committed:
prettier

abovedave committed:
add ignore

abovedave committed:
retry ignore

abovedave committed:
add missing curly

abovedave committed:
remove ignore

abovedave committed:
add paths to tsconfig

abovedave committed:
try another way

abovedave commented on package.json:
```diff
@@ -33,33 +33,33 @@
     "scripts": {
         "start": "npm run build",
-        "build": "node --no-warnings --loader ts-node/esm ./scripts/build/index.ts",
+        "build": "tsc --noEmit && node --no-warnings --loader ts-node/esm ./scripts/build/index.ts",
```
[Got this from here](https://stackoverflow.com/a/77993035/11399923), otherwise it was failing without any errors.

dancormier replied:
Since you've added `"noEmit": true` to `tsconfig.json`, I think this shouldn't be necessary.

abovedave commented on scripts/build/write-js.ts:
```diff
@@ -1,7 +1,7 @@
 import { promises as fs } from "fs";
 import { rollup } from "rollup";
-import packageJson from "../../package.json" assert { type: "json" };
+import packageJson from "../../package.json" with { type: "json" };
```
This changed in [typescript 5.3](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-3.html) but was legacy

abovedave commented on src/js/global.d.ts:
```diff
@@ -1,4 +1,4 @@
 declare module "@stackoverflow/stacks-icons" {
-    export * as Icons from "@stackoverflow/stacks-icons/icons";
-    export * as Spots from "@stackoverflow/stacks-icons/spots";
+    export namespace Icons {}
+    export namespace Spots {}
```
Got a bit stuck here with an error of the package not being found before build, not sure if this breaks anything or not.

The errors were:

```
Cannot find module '@stackoverflow/stacks-icons/icons' or its corresponding type declarations.
Cannot find module '@stackoverflow/stacks-icons/spots' or its corresponding type declarations.
```

dancormier replied:
Hmm that's odd. I can't reproduce that locally and the original exports make the icon and spot types available when developing around the codebase.

Is this reported from a regular `npm start`/`npm run build`?

giamir replied:
Just want to recommend caution with changing those declaration files. They could cause unexpected type errors for consumers down the line. If changes need to be made here I would recommend to manually test them in Core before merging to see how the TS compiler reacts there. 🙂

abovedave replied:
You can replicate by deleting your `dist` folder to simulate a fresh build.

I've figured it out by using `@ts-expect-error`, which is what we use in the other files in this folder. Should be good now!

abovedave replied:
Ok small update, that works on the CI, but locally it means on the **next** run it throws an error because it is _expecting the error_, so `@ts-ignore` looks the best bet, but I had to push an update to the link rules to allow it with a description.

Alternative would maybe be just be a `@ts-nocheck` in this file?

dancormier commented on package.json:
```diff
@@ -33,33 +33,33 @@
         "preview": "http-server ./preview -o",
         "prepublishOnly": "npm start",
-        "version": "node --no-warnings --loader ts-node/esm ./scripts/sync-dotnet-version.ts",
+        "version": "tsc --noEmit && node --no-warnings --loader ts-node/esm ./scripts/sync-dotnet-version.ts",
```
`tsc --noEmit` shouldn't be necessary here either

dancormier commented on tsconfig.json:
```diff
@@ -18,8 +18,10 @@
         "checkJs": true,
-        "outDir": "./tmp",
-        "noEmitOnError": true
+        "noEmit": true,
+        "noEmitOnError": true,
```
I believe `"noEmitOnError": true` is made redundant by the more broad `"noEmit": true`

```suggestion
        "noEmit": true,
```

dancormier commented on scripts/build/write-manifests.ts:
```diff
@@ -54,13 +53,18 @@ export async function writeManifests(
-        ],
-        paths.build("index.d.ts")
+    const p2 = await fs.writeFile(
+        paths.build("index.d.ts"),
+        (
+            await Promise.all(
+                [
+                    paths.src("js/global.d.ts"),
+                    paths.build("icons.d.ts"),
+                    paths.build("spots.d.ts"),
+                ].map((f) => fs.readFile(f, "utf8"))
+            )
+        ).join("\n\n"),
+        "utf8"
```
👍 LGTM!

dancormier reviewed (changes requested):
Thanks for this housekeeping PR @abovedave!

> Upgrades all dependencies to latest versions, including typescript.

💯 Thank you!

>     Adds an .nvmrc file as we've done in other packages.

👍 This seems reasonable enough. I dug around in the `stacks-*` repos and found that we don't include an `.nvmrc` file in any of them. Makes me wonder if it's worth adding just for consistency 🤔 

>     Removes the old [concat](https://www.npmjs.com/package/concat) package which was causing ts errors in favour of a native solution.

One fewer dependencies. Nice!

>     Couple of changes to tsconfig.json to resolve build errors.

I made some notes about this and the changes to `package.json` scripts. Just a little cleanup.

abovedave committed:
tidy use of noEmit

abovedave committed:
use @ts-expect-error to silence build error

abovedave committed:
swap `@ts-expect-error` for `@ts-ignore`

After first build there is no longer an error produced here, so ignore is probably a better choice

abovedave committed:
allow `@ts-ignore` with comment

abovedave committed:
run format

dancormier reviewed (approved):
This looks reasonable to me. Thanks @abovedave!
